### ISSUE TITLE
Evaluate Google Cloud Build for Image building

### ISSUE THREAD

**Issue by Ressetkk** [kyma-project/test-infra#4935]
Right now we are building our Docker images inside Prow cluster. There would be nothing bad with that approach, but it requires us to maintain jobs and scripts for all images across the organization. This also requires running most of our pipelines in privileged mode because those jobs require Docker daemon.

GCP offers Google Cloud Build service which allows using Google cloud for building the images without the need of Docker in Prow. This will harden the security concerns a bit.

This issue aims to find a solution for building images that are not considered as binaries, eg. kyma-integration image, various buildpacks, base image for rootless builds and other images that need to run integrations. We should aim for more automated approach and more accessible for developers, who are more familiar with Prow. More in #4755 

AC for image builder:

- [ ] We need to get rid of privileged containers in test-infra (SLC-29).
- [ ] Ideally works with all images we build on Prow.
- [ ] Must allow us to secure write access to GCR. Credentials needed to push images must be accessible only for controlled group of user and bot accounts (SLC-29). https://github.tools.sap/kyma/backlog/issues/2430
- [ ] The way how developers can setup build for their images should not change too much. They still must be able to add some definition/manifest to github repo, setup prowjob which use this definition/manifest and have their image build when conditions for prowjob are met.
- [ ] Must allow us to hook in a building process on different stages (security scans, image signing, automation).
- [ ] Should allow us to use caches for builds.
- [ ] Should allow to build multi-architecture images (ARM, AMD64).

**Comment by Ressetkk** [comment]
Time to unfreeze this!
/remove-lifecycle frozen
/assign

**Comment by Ressetkk** [comment]
So, upon further investigation I have managed to achieve similar building approach with Google Cloud Build, as in our current approach. This is working basically with leveraging build processes to the external infrastructure that handles their own Docker environment and every event triggers a predefined trigger on GCP. Those triggers work based on GitHub events.
![image](https://user-images.githubusercontent.com/23667947/170235620-c152c4cb-5d26-4e1b-a227-331e6d7714a2.png)
For running workloads GitHub App is installed and authorised to use specific, or all repositories.
For configuring what is supposed to run, we have 3 things to use:
* `cloudbuild.yaml` file - uses custom YAML file that defines build image, and specific steps that run in a cloud build. Gives the most flexibility
* Dockerfiles - Fetches the docker file and builds them against defined context (docker directory). Not possible to add custom workloads and doesn't give flexible configuration. 
* Buildpacks
The 4th option is to auto-detect which of the following 3 is used.
For the evaluation I've used the first option, as it gives the most configuration options.

I've picked 2 images: bootstrap and buildpack-golang.
After linking my repo to the GCP project, creating [cloudbuild.yaml file](https://github.com/Ressetkk/test-infra/blob/cloud-build-test/images/bootstrap/cloudbuild.yaml) and creating a trigger that invokes only when specific files are modified. It gave a nice status check pushed to the GitHub and shows logs quite well. Additionally substitutions can be used to re-use the following configuration in other triggers, like in the PR, of which example you can see [here](https://github.com/Ressetkk/test-infra/pull/1).

Wha pros I see so far:
* Building images completely independent from Prow. No secrets are needed, external application manages everything
* Relatively easy usage of environment substitiution
* No need to use Docker daemon and sophisticated building pipelines to build Docker images

Cons:
* Not very easy way of using bash commands for filling up env variables - for example to use  `$(date +v%Y%m%d)` we have to use bash as an entry point of an image
* There are problems with bash default value substitutions `${var:-default}`. It doesn't detect assigned value for some reason. Further investigation required.
* Building multi-arch builds are not so straightforward. The only example I've found so far suggest using terraform to provision separate environments to build the images.

Another option is to use gcloud image in the ProwJob and run the GCB workload from Prow. That should give the ability to run more sophisticated env substitutions and will provide logs to the ProwJob itself, instead of using external logging solution, which might be blocked for external contributors. This would require more strict security configurations for secrets (and secret obscurity).

I would see GCB as good infra for building the testimages we use as entrypoints in ProwJobs. For example kyma-integration, bootstrap, various buildpacks etc. However when it comes to releasing the images of the binaries, I would aim for other solution, like provide a buildpack image that will include all tooling required for building images. However, we can also use GCB for that for managing triggers, but with buildpacks option.

**Comment by Ressetkk** [comment]
I've also tested approach, where we have one big `cloudbuild.yaml` file, and all images are built concurrently. This approach would allow us to prepare only static set of triggers which would build all images one by one, and if needed we could re-use cache, if some images depend on each other.
* That would be ideal for approach "code and build" - no need to set up infra to have your image built, just edit a file on a repo
* That would rebuild all image every time one gets changed. A little bit overhead, but assume all images are synchronised and up-to-date with each other.
* This would require creating more complex cloudbuild.yaml file and maintain it

For me both approaches work great, however since we have automation that automatically bumps up our images, I would go with the second approach. We could also drop "PR images" since we don't really use them.

**This only applies for the test-infra images! This doesn't resolve the problem of building components images, but results can be used to find a suitable solution for them!**

**Comment by dekiel** [comment]
@Ressetkk this issue is missing detailed description of requirements and problems we would like to solve by this implementation. For me one of requirements of this issue is to solve privileged containers issue on Prow. With such requirement we can't evaluate solution for images in test-infra only. Once we will have requirements defined we can discuss about proposed solution.

For me this issue should solve or require following:
1. We need to get rid of privileged containers in test-infra.
2. Must work with all images we build on Prow.
3. Must allow us to secure write access to GCR. Credentials needed to push images must be accessible only for controlled group of user and bot accounts.
4. The way how developers can setup build for their images should not change too much. They still must be able to add some definition/manifest to github repo, setup prowjob which use this definition/manifest and have their image build when conditions for prowjob are meet.
5. Must allow us to hook in a building process on different stages.
6. Should allow us to use caches for builds.

**Comment by Ressetkk** [comment]
@dekiel it's a broader topic and most likely will require more work, than evaluating different build infrastructure. This issue only tackles the problem of evaluating Google Cloud Build to use it for building our images outside of Prow so we can remove the need for privileged containers.

**Comment by Ressetkk** [comment]
Our images are all images that we have under `prow/images` directory. Those directly work on Prow as bases for ProwJobs. To remove usage of Docker daemon in Prow and still build them - we need to find a solution. This is the one, for which I've created demo and shared my insights. It's a part of a bigger step to complete remove the usage of docker daemon, and is not a go-to solution for everyone. I've updated the Description of #4755 so they reflect the requirements with required and optional ones.

**Comment by Ressetkk** [comment]
Multi-arch builds are shown [here](https://cloud.google.com/architecture/building-multi-architecture-container-images-iot-devices-tutorial).

I've tested using GCB from console on test-infra images (bootstrap, buildpack-golang (it will work the same way in Prow) and on one of the Kyma components. Where this configuration works for test-infra, the components would be a tad bit complex to introduce. Because we have not standard configurations, where we keep Dockerfiles somewhere inside the subfolders of build context, see components/application-broker for example, each component job requires it's own `cloudbuild.yaml` file with the configuration.

Maybe we could also try to adopt https://github.com/kubernetes/test-infra/tree/master/images/builder. It's a wrapper for `gcloud builds submit`, or write our own tool that works around it.

Substituting `docker build` with `gcloud builds submit` and cloudbuild.yaml config files should be enough. We have to rewrite all pipelines though.

**Comment by k15r** [comment]
Regarding requirements:

**Comment by Ressetkk** [comment]
GCB is just build infrastructure. We can use it for building images the same way how we build them with DinD. With that we can meet requirements for SLC-29 regarding not using Docker daemon in jobs.

Criteria we meet:
✅ We can remove Docker in Docker for building images
✅ It will work with all images that we build on Prow right now - just need proper Dockerfile and cloudbuild.yaml file
✅ Jobs will be triggered from Prow so the configuration for the devs will not change too much (still cloudbuild.yaml file will be required)
✅ We will be able to hook into the building process, be it with the Makefile stages or in the building tool itself (will be written in Go)
✅ We will be able to use caches if we decide to build multiple images at once (defined in one cloudbuild.yaml file)
✅ Developers will be able to build their local images with Docker

🟥 We will not be able to run multi-arch builds unless we provide separate configurations for building infrastructure platforms (see my comment above)
🟥 For accessing the build infrastructure we use secret that has rights to executing Google Cloud Builds inside Prow. Then the GCB will push image to the registry. Unless we restrict access to testing ProwJobs on PRs, we will never meet criteria for restricting registry access. This problem should have it's own issue.

**Comment by dekiel** [comment]
@Ressetkk 
>For accessing the build infrastructure we use secret that has rights to executing Google Cloud Builds inside Prow. Then the GCB will push image to the registry. Unless we restrict access to testing ProwJobs on PRs, we will never meet criteria for restricting registry access. This problem should have it's own issue.

This problem has it's own issue. Check [slc-29] Restrict write access to OCI registries for CI/CD technical users. internal issue.

We should not start any implementation before we know last requirement will be possible to meet or we must ask for exception.

**Comment by dekiel** [comment]
@Ressetkk do you have some draft PR or dev setup where I can read `cloudbuild.yaml` files, see directory structure, read prowjob spec and code it execture, check GCB workflow configuration.

I read k8s builder docs. It looks that we could use it, maybe it will require some extend. However I need to understand a whole setup to comment.
Regarding securing access to credentials, a usage of google meta server looks like a solution for us.

**Comment by Ressetkk** [comment]
I haven't created ProwJobs since I have to prepare Prow to use workload identity. You can check the cloudbuild.yaml files that I linked in the comments above. I don't really know if it's a solution - we have to create k8s SA -> GCP SA bindings and those also can be changed in ProwJob definition. This will only remove the requirement of attaching GCP keys as secrets in pods.

 Unless we restrict access to k8s SA, eg. enforce pjtester jobs to use restricted service account regardless the configuration, we will not achieve this requirement.

**Comment by dekiel** [comment]
@Ressetkk files you linked are for scenario with one shared cloudbild.yaml and for scenario with triggers defined directly on GCB. As far I understand the last scenario which meets most of our requirements is different. Is one of cloudbuild.yaml files you linked the one used in last scenario?

**Comment by Ressetkk** [comment]
@dekiel Yes, it is required. The way the build triggers only gets changed - you use gcloud command to run the workload.

**Comment by dekiel** [comment]
@Ressetkk which one?

**Comment by Ressetkk** [comment]
Closing in favor of #5637

/close
